<a href="https://colab.research.google.com/github/240609/MLF/blob/main/RecurrentNN/TheMovieReview.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**2.0 - Import libraries + 2.1 - Load dataset**

In [1]:
#from keras.preprocessing.text import Tokenizer
import keras
import numpy as np
from keras.models import Sequential
from keras.layers import LSTM, Dense, Embedding, SimpleRNN, GRU
from keras.preprocessing.sequence import pad_sequences
import matplotlib.pyplot as plt
from keras.optimizers import Adam

size = 1000
(X_train, y_train), (X_test, y_test) = keras.datasets.imdb.load_data(num_words=size)
print(X_train.shape)
print(X_test.shape)

17464789/17464789 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step
(25000,)
(25000,)


**2.2 - Dataset examination**

In [2]:
word_index = keras.datasets.imdb.get_word_index()
id_to_word = {id_ + 3:word for word, id_ in word_index.items()}
for id_, token in enumerate(("<pad>", "<sos>", "<unk>")):
  id_to_word[id_] = token

single_sequence = " ".join([id_to_word[id_] for id_ in X_train[0][:]])

single_sequence


1641221/1641221 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step


"<sos> this film was just brilliant casting <unk> <unk> story direction <unk> really <unk> the part they played and you could just imagine being there robert <unk> is an amazing actor and now the same being director <unk> father came from the same <unk> <unk> as myself so i loved the fact there was a real <unk> with this film the <unk> <unk> throughout the film were great it was just brilliant so much that i <unk> the film as soon as it was released for <unk> and would recommend it to everyone to watch and the <unk> <unk> was amazing really <unk> at the end it was so sad and you know what they say if you <unk> at a film it must have been good and this definitely was also <unk> to the two little <unk> that played the <unk> of <unk> and paul they were just brilliant children are often left out of the <unk> <unk> i think because the stars that play them all <unk> up are such a big <unk> for the whole film but these children are amazing and should be <unk> for what they have done don't you

In [3]:
max_length = 150  # Define the maximum length of a review
X_train = keras.preprocessing.sequence.pad_sequences(X_train, maxlen=max_length, padding='post', truncating='post', value=0)
X_test = keras.preprocessing.sequence.pad_sequences(X_test, maxlen=max_length, padding='post', truncating='post', value=0)

embed_size = 128
vocabulary_size = 1000




**2.3 RNN**

In [ ]:
model = keras.models.Sequential([
    keras.layers.Embedding(vocabulary_size, embed_size,
                           input_shape=[None]),
    keras.layers.SimpleRNN(128, return_sequences=True),
    keras.layers.SimpleRNN(128),
    keras.layers.Dense(1, activation='sigmoid')
    ])

optimizer = Adam(learning_rate = 0.001)

model.compile(loss="binary_crossentropy", optimizer=optimizer, metrics=["accuracy"])

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/embedding.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [ ]:
history = model.fit(X_train, y_train, epochs=5, validation_split=0.2)

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print(f'Test accuracy: ', score[1]*100, "%")

Epoch 1/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 97s 151ms/step - accuracy: 0.5011 - loss: 0.7235 - val_accuracy: 0.4938 - val_loss: 0.6940
Epoch 2/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 135s 140ms/step - accuracy: 0.5004 - loss: 0.6967 - val_accuracy: 0.4938 - val_loss: 0.6933
Epoch 3/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 146s 147ms/step - accuracy: 0.4977 - loss: 0.6960 - val_accuracy: 0.5062 - val_loss: 0.6957
Epoch 4/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 142s 147ms/step - accuracy: 0.4978 - loss: 0.6961 - val_accuracy: 0.4938 - val_loss: 0.6934
Epoch 5/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 135s 137ms/step - accuracy: 0.4935 - loss: 0.6948 - val_accuracy: 0.5062 - val_loss: 0.6967
Test loss: 0.6978574991226196
Test accuracy:  50.0 %


**2.4 LSTM**

In [ ]:
model = keras.models.Sequential([
    keras.layers.Embedding(vocabulary_size, embed_size,
                           input_shape=[None]),
    keras.layers.LSTM(128, return_sequences=True),
    keras.layers.LSTM(128),
    keras.layers.Dense(1, activation='sigmoid')
    ])

optimizer = Adam(learning_rate = 0.001)

model.compile(loss="binary_crossentropy", optimizer=optimizer, metrics=["accuracy"])

In [ ]:
history = model.fit(X_train, y_train, epochs=5, validation_split=0.2)

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print(f'Test accuracy: ', score[1]*100, "%")

Epoch 1/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 356s 563ms/step - accuracy: 0.5410 - loss: 0.6834 - val_accuracy: 0.5506 - val_loss: 0.6807
Epoch 2/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 391s 577ms/step - accuracy: 0.6041 - loss: 0.6518 - val_accuracy: 0.7794 - val_loss: 0.4892
Epoch 3/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 370s 558ms/step - accuracy: 0.8006 - loss: 0.4446 - val_accuracy: 0.8208 - val_loss: 0.4009
Epoch 4/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 347s 556ms/step - accuracy: 0.8359 - loss: 0.3770 - val_accuracy: 0.8278 - val_loss: 0.3876
Epoch 5/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 361s 578ms/step - accuracy: 0.8449 - loss: 0.3497 - val_accuracy: 0.8360 - val_loss: 0.3834
Test loss: 0.37231186032295227
Test accuracy:  83.51600170135498 %


**2.5 GRU**

In [4]:
model = keras.models.Sequential([
    keras.layers.Embedding(vocabulary_size, embed_size,
                           input_shape=[None]),
    keras.layers.GRU(128, return_sequences=True),
    keras.layers.GRU(128),
    keras.layers.Dense(1, activation='sigmoid')
    ])

optimizer = Adam(learning_rate = 0.001)

model.compile(loss="binary_crossentropy", optimizer=optimizer, metrics=["accuracy"])

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/embedding.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [5]:
history = model.fit(X_train, y_train, epochs=5, validation_split=0.2)

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print(f'Test accuracy: ', score[1]*100, "%")

Epoch 1/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 275s 435ms/step - accuracy: 0.5177 - loss: 0.6906 - val_accuracy: 0.7214 - val_loss: 0.5585
Epoch 2/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 319s 431ms/step - accuracy: 0.7715 - loss: 0.4844 - val_accuracy: 0.8286 - val_loss: 0.3860
Epoch 3/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 323s 432ms/step - accuracy: 0.8416 - loss: 0.3514 - val_accuracy: 0.8420 - val_loss: 0.3696
Epoch 4/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 324s 436ms/step - accuracy: 0.8608 - loss: 0.3213 - val_accuracy: 0.8330 - val_loss: 0.3783
Epoch 5/5
625/625 ━━━━━━━━━━━━━━━━━━━━ 317s 428ms/step - accuracy: 0.8747 - loss: 0.2899 - val_accuracy: 0.8310 - val_loss: 0.3864
Test loss: 0.3721016049385071
Test accuracy:  83.55600237846375 %


**2.6 Comparation**

**SimpleRNN**

Test loss: 0.6978574991226196

Test accuracy:  50.0 %


**LSTM**

Test loss: 0.37231186032295227

Test accuracy:  83.51600170135498 %



**GRU**

Test loss: 0.3721016049385071

Test accuracy:  83.55600237846375 %